lib/EarlyFunctionAnalysis/DetectABI.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/Pass.h"
+
+#include "revng/ADT/Queue.h"
+#include "revng/BasicAnalyses/GeneratedCodeBasicInfo.h"
+#include "revng/EarlyFunctionAnalysis/ABIAnalysis.h"
+#include "revng/EarlyFunctionAnalysis/CFGAnalyzer.h"
+#include "revng/EarlyFunctionAnalysis/CallGraph.h"
+#include "revng/EarlyFunctionAnalysis/FunctionSummaryOracle.h"
+
+namespace efa {
+
+using BasicBlockQueue = UniquedQueue<const BasicBlockNode *>;
+
+class DetectABI {
+private:
+  using CSVSet = std::set<llvm::GlobalVariable *>;
+
+private:
+  llvm::Module &M;
+  llvm::LLVMContext &Context;
+  GeneratedCodeBasicInfo &GCBI;
+  TupleTree<model::Binary> &Binary;
+  FunctionSummaryOracle &Oracle;
+  CFGAnalyzer &Analyzer;
+
+  BasicBlockQueue EntrypointsQueue;
+
+  CallGraph ApproximateCallGraph;
+
+public:
+  DetectABI(llvm::Module &M,
+            GeneratedCodeBasicInfo &GCBI,
+            TupleTree<model::Binary> &Binary,
+            FunctionSummaryOracle &Oracle,
+            CFGAnalyzer &Analyzer) :
+    M(M),
+    Context(M.getContext()),
+    GCBI(GCBI),
+    Binary(Binary),
+    Oracle(Oracle),
+    Analyzer(Analyzer) {}
+
+public:
+  void run();
+  ABIAnalyses::ABIAnalysesResults analyzeABI(llvm::BasicBlock *Entry);
+  ABIAnalyses::ABIAnalysesResults
+  analyzeABI(MetaAddress Addr) { // TODO: Tidy this
```

Comment:
Use `// WIP:` for comments that should not be merged. `check-conventions` will complain about them.